### ISSUE TITLE
`MutationFunctionOptions.optimisticResponse` type is too restrictive in typescript 3.9

### ISSUE THREAD

**Issue by seansfkelley** [apollographql/apollo-client#6390]
**Intended outcome:** the following code compiles with Typescript 3.9.

```ts
import { OperationVariables, Context, MutationFunctionOptions } from "@apollo/react-common";
import { useMutation, MutationHookOptions } from "@apollo/react-hooks";

export interface MyFunctionContext<
  TData = any,
  TVariables = OperationVariables
> {
  extraConfig: MyConfigType<TData, TVariables>;
}

export type MyMutationFunctionOptions<TData, TVariables> = Omit<
  MutationFunctionOptions<TData, TVariables>,
  "context" | "optimisticResponse"
> & {
  context: MyFunctionContext<TData, TVariables> & Context;
  optimisticResponse: TData | ((vars: TVariables) => TData);
};

export const useMyMutation = <TData = any, TVariables = OperationVariables>(
  mutation: DocumentNode,
  options?: MutationHookOptions<TData, TVariables>,
) => {
  const [mutationFunc, mutationResult] = useMutation(mutation, options);
  const newMutationFunc = useCallback(
    (options: MyMutationFunctionOptions<TData, TVariables>) => mutationFunc(options),
    [mutationFunc],
  );
  return [newMutationFunc, mutationResult] as const;
};

```

**Actual outcome:** tsc rejects the signature of `optimisticResponse`, though it did compile on Typescript 3.7.

```
TS2345: Argument of type 'MyMutationFunctionOptions<TData, TVariables>' is not assignable to parameter of type 'MutationFunctionOptions<TData, TVariables> | undefined'.                                                                                                                                                                  
  Type 'MyMutationFunctionOptions<TData, TVariables>' is not assignable to type 'MutationFunctionOptions<TData, TVariables>'.                                                                       
    Types of property 'optimisticResponse' are incompatible.                                                                                                                                                     
      Type 'TData | ((vars: TVariables) => TData)' is not assignable to type 'TData | ((vars: {} | TVariables) => TData) | undefined'.                                                                           
        Type '(vars: TVariables) => TData' is not assignable to type 'TData | ((vars: {} | TVariables) => TData) | undefined'.                                                                                   
          Type '(vars: TVariables) => TData' is not assignable to type '(vars: {} | TVariables) => TData'.
            Types of parameters 'vars' and 'vars' are incompatible.
              Type '{} | TVariables' is not assignable to type 'TVariables'.                            
                'TVariables' could be instantiated with an arbitrary type which could be unrelated to '{} | TVariables'.
                  Type '{}' is not assignable to type 'TVariables'.                                     
                    'TVariables' could be instantiated with an arbitrary type which could be unrelated to '{}'.
                                                                                                        
96     (options: MyMutationFunctionOptions<TData, TVariables>) => mutationFunc(options),
```

**How to reproduce the issue:**

This can be boiled down to the `| {}` union in `optimisticResponse`.

```ts
type Fn<T> = (vars: T | {}) => unknown;

interface Variables {
  foo: string;
}

const myFunction: (vars: Variables) => unknown = null as any;

const fails: Fn<Variables> = myFunction;
```

[Playground Link](https://www.typescriptlang.org/play/#code/C4TwDgpgBAYgdgHgCoD4oF4oAoBuBDAJwGcAuKJKAHygG8BfASgzQFc4BrOAewHc4BuAFCCAlnGAQCAMzwBjaADVCIvACMANhCK1BUKFK5cyRYATEBzIXWGyucE1AC2IGG1nARdsrkKkoSszVNIiZ0Vg5uPgwoOBZ1dSg8bTw4ECFBW3tgfTwRdT94BACVDS00TGdXOHdPASA)

I don't understand what purpose `| {}` is addressing when the default type is `TVariables = OperationVariables` and `type OperationVariables = Record<string, any>`, so I think the fix is to just remove it. A `TVariables extends object` constraint would probably make sense too.

**Versions**

- `apollo-client@2.6.4`
- `@apollo/react-common@3.1.1`

Though the problematic `| {}` is [still on `@apollo/client`'s unpublished master](https://github.com/apollographql/apollo-client/blob/master/src/react/types/types.ts#L179).

**Comment by scott-rc** [comment]
The change appears to have been made via this [PR](https://github.com/apollographql/react-apollo/pull/3257/files) last year. Still not sure why the `| {}` was added though... I'd prefer not to add `@ts-ignore` all the time.

**Comment by seansfkelley** [comment]
Nice find @scott-rc. It looks like it might even have been a mistake, since in that PR one version of `optimisticResponse` has `| {}` and the other does not, though before the PR they used to be identical. Maybe a simple leftover from an earlier implementation?

**Comment by seansfkelley** [comment]
Closed by #6643!
